src/librustc_error_codes/lib.rs

```diff
@@ -0,0 +1,26 @@
+//! This library is used to gather all error codes into one place. The goal
+//! being to make their maintenance easier.
+
+#[macro_export]
+macro_rules! register_diagnostics {
+    ($($ecode:ident: $message:expr,)*) => (
+        $crate::register_diagnostics!{$($ecode:$message,)* ;}
+    );
+
+    ($($ecode:ident: $message:expr,)* ; $($code:ident,)*) => (
+        pub static DIAGNOSTICS: &[(&str, &str)] = &[
+            $( (stringify!($ecode), $message), )*
+        ];
+
+        $(
+            pub const $ecode: &str = $message;
+        )*
+        $(
+            pub const $code: () = ();
+        )*
+    )
+}
+
+mod error_codes;
+
+pub use error_codes::*;
```

Comment:
Could just export `DIAGNOSTICS` now that it is the only item of interest; alternatively, since it is the only item, we could inline the whole `error_codes` file into `lib.rs`.

Reply:
I'd prefer not. The file is big enough on its own so let's keep what's error code specific outside if possible.

Reply:
It becomes small after using `include!(...)`s. ;)

Reply:
(unless the error codes long explanation are moved into files.

Reply:
Even after the change, I still think it's better to keep it in its own file. :)